libnmstate/nm/connection.py

```diff
@@ -294,6 +300,28 @@ def waitfor_active_connection_async(self, ac):
                 "state-changed", self._waitfor_device_state_change_callback, ac
             )
         )
+        # As we also depend on IP4_READY flag which does not trigger
+        # 'state-change' signal, hence do a repeated recheck without timeout.
+        ac.state_check_running = True
+        nmclient.GLib.timeout_add_seconds(
+            ConnectionProfile._ACTIVATION_CHECK_INTERNAL,
+            self._check_device_state,
+            ac,
+        )
+        logging.debug("Activation state check started")
+
+    def _check_device_state(self, ac):
+        if ac.state_check_running and (
```

Comment:
Using `ac.state_check_running` is also a bit ugly, seems this is controlled only from this object. I wander if we can use some other existing state to do the same... like the "handlers" existence. At the minimum, to use some generic state and not something specific to "checking".